version/main.go

```diff
@@ -0,0 +1,477 @@
+
+import (
+	"context"
+	"crypto/sha1"
+	"dagger/version/internal/dagger"
+	"encoding/hex"
+	"fmt"
+	"path/filepath"
+	"slices"
+	"strconv"
+	"strings"
+
+	"golang.org/x/mod/semver"
+)
+
+func New(
+	// A directory containing all the inputs of the artifact to be versioned.
+	// An input is any file that changes the artifact if it changes.
+	// This directory is used to compute a digest. If any input changes, the digest changes.
+	// - To avoid false positives, only include actual inputs
+	// - To avoid false negatives, include *all* inputs
+	// +optional
+	// +defaultPath="/"
+	// +ignore=["**_test.go", "**/.git*", "**/.venv", "**/.dagger", ".*", "bin", "**/node_modules", "**/testdata", "**.changes", "docs", "helm", "release", "version", "modules", "*.md", "LICENSE", "NOTICE", "hack"]
+	inputs *dagger.Directory,
+	// +optional
+	// +defaultPath="/"
+	// +ignore=["*", "!.git"]
+	gitDir *dagger.Directory,
+	// .changes file used to extract version information
+	// +optional
+	// +defaultPath="/"
+	// +ignore=["*", "!.changes/*"]
+	changes *dagger.Directory,
+) Version {
+	return Version{
+		// FIXME: upload the whole git dir is inefficient.
+		// We can stop doing it once this is shipped:  https://github.com/dagger/dagger/issues/8520
+		GitDir:  gitDir,
+		Inputs:  inputs,
+		Changes: changes,
+	}
+}
+
+type Version struct {
+	GitDir *dagger.Directory
+	Inputs *dagger.Directory
+
+	Changes *dagger.Directory
+}
+
+// Return whether the current version is a development version or not
+func (v Version) Dev(ctx context.Context) (bool, error) {
+	tag, err := v.VersionTag(ctx)
+	if err != nil {
+		return false, err
+	}
+	// The current implementation is:
+	// - If the current commit is semver-tagged, dev=false
+	// - Otherwise, dev=true
+	// FIXME: this is a flawed implementation..
+	return (tag == ""), nil
+}
+
+// Generate a version string from the current context
+// - If the current commit is a semver tag, the version is that tag
+// - If no git information is available, the version is "<next-version>-<diget>"
+// - In all other cases, the version is "<next-version>-<commit>-<digest>", where:
+//   - <next-version> is inferred from .changes/
+//   - <digest> is computed from the inputs directory
+func (v Version) Version(ctx context.Context) (string, error) {
+	if v.gitDirExists(ctx) {
+		tag, err := v.VersionTag(ctx)
+		if err != nil {
+			return "", err
+		}
+		if tag != "" {
+			// FIXME: we don't handle dirty checkout of a semvar tag checkout
+			return tag, nil
+		}
+	}
+	next, err := v.NextVersion(ctx)
+	if err != nil {
+		return "", err
+	}
+	digest, err := v.InputsDigest(ctx)
+	if err != nil {
+		return "", err
+	}
+	if !v.gitDirExists(ctx) {
+		return fmt.Sprintf("%s-%s", next, digest), nil
+	}
+	commit, err := v.Commit(ctx)
+	if err != nil {
+		return "", err
+	}
+	// FIXME we don't differentiate between clean and dirty checkout:
+	// instead we always add the commit + digest
+	return fmt.Sprintf("%s-%s-%s", next, commit, digest), nil
```

Comment:
We need to make sure that "dev" builds get a proper "dev" entry in their tag - this is load-bearing: https://github.com/dagger/dagger/blob/42e8daf846dffe91476679d5cf0d3955630c9171/engine/version.go#L133-L138

Also, note on terminology here - there are 3 cases:
- Release builds (have an associated tag)
- Main builds (have an associated commit, but no tag)
	- These are still technically released and uploaded to the registry and s3
- Dev builds (made locally, have no associated commit or tag)
	- This isn't the same as a `main` build
I think we need to keep this terminology, see the above load-bearing code snippets.

That said, I'm not 100% sure how we do this. Ideally dev should be set only on dirty checkouts, but I'm not quite sure how we compute that. The only thing I can think of is to import *everything*, so we can run a `git status`, but that's *so* expensive. We really need contextual git #8520, but not sure what a temporary workaround looks like.

*Maybe* we could have the main dagger module keep track of a `--dev` flag? Then when building locally, you have to make sure to set a `--dev` flag (we can set this automagically in `./hack/dev` that most of us use).